README.md
# hauth: authentication, authorization and accounting module for web apps with PG database

This `hauth` package provides secured client password-based authentication so that express-based web applications can reuse without rewriting most frequently used authentication flows.

## Introduction

This module is aimed at authenticating web users as well as devices.
The use of this module assumes that:

- the app is powered by [express](https://www.npmjs.com/package/express)
- the app can connect to a [Postgres](https://www.postgresql.org/about/) database, and has right to create tables on that database

## Features

### Password Authentication

If a client makes a request to a path for which the app expects authentication, the app sends an `HTTP` response with a `401` status code, a reason phrase indicating an authentication error, and a `WWW-Authenticate` header.

The client can then submit login credentials in either request body, or request header using `HTTP Basic Authentication` with `Authorization` header:

```raw
Authorization: Basic login:password
```

### Role-based Access Control

Hauth manages access rules based on URLs path (without query string) and on roles. Each user can be assigned one role. Access rules look like :

```cfg
'/node_modules': 'skip',    // disable access control => no authentication required
'\.css$':        'skip',    // as same - URLs can be expressed with regex
'/app':          'allow',   // allowed to any authenticated user
'/reserved':     ['admin'], // access granted only to users with role 'admin'
'/':             'deny',    // forbid everything which is not allowed
```

By default, access is denied if no access rule matches.

### Accessing User Data

The app can retrieve user data in `req.user` field.

### Custom Login and Forbidden Pages

In case of missing credentials or access denied, Hauth will respond with 401 or 403 ; else, the response is managed by the application.

Besides, it is possible to manage custom 401 or 403 response, for example in order to provide an authentication form or a "Forbidden" page.

### Session Storage

- User data are stored in a [JSON Web Token](https://en.wikipedia.org/wiki/JSON_Web_Token) (JWT), so the session data is stored only on client's side, and not in the database.
- Ability to configure JWT key, algorithm and expiration time
- It is advised to set a JWT key: indeed, if not set a random key is computed at each server startup, so after each startup, JWT previously delivered are not valid anymore and users must reauthenticate.

### Passwords Stored Hashed

Salted passwords are hashed using [bcrypt](https://www.npmjs.com/package/bcrypt),
an implementation of OpenBSD Blowfish password hashing algorithm.

The computation cost of `bcrypt` algorithm is parametized, which means it can be increased to adapt to future hashing power improvements in case of brute-force attacks. This most important property of `bcrypt` makes it the recommeded hashing algorithm to hash user's password.

### Force Password Change at Next Login

Intended for devices, in case we need to change password (which the device uses to authenticate itself with the app) for security reasons.
The module can generate the `Next-Password`, and the app will include this `Next-Password` in the response header after a sucessfull authentication, from the next time, the device must use this `Next-Password` to authenticate.

### Account Management

Users can be added, deleted, modified (updating its role, name, changing password, etc.) easily via the exposed functions.

### Automatic Account Generation

User accounts can be generated and their details are filled properly in the `hauth_user` and `hauth_role` by defining in the configuration.

### Default Accounts

The default user accounts are created only at the first use of the module, right after the creation of `hauth_user` and `hauth_role` tables, so we always have default accounts ready to log in (to avoid the situation where there is no account). These accounts can be deleted after.

## Usage

Create an express app, `require` the module `hauth`, `init` it with `db` and `config` parameters, and enjoy!

```js
const express = require('express');
const app = express();

const pg = require('pg');
const db = new pg.Pool( /* PG params */ );
const config = { /* cf doc below */ };
const hauth = require('@horanet/hauth');
hauth.init(config, db);

app.use('/hauth/login', hauth.getCookie);
app.use('/hauth/logout', hauth.delCookie);
app.use('/', hauth.control);

app.use( /* app code */ );
app.listen(port, () => { /* ... */ });
```

For more features, look at the example provided.

### PG tips

Assume the database parameters are as follows:

```cfg
host: 'localhost',
port: 5432,
database: 'test',
user: 'testuser',
password: 'testpass',
```

one can create the database using `psql`:

```psql
sudo -u postgres psql
postgres=# CREATE DATABASE test;
postgres=# CREATE USER testuser WITH ENCRYPTED PASSWORD 'testpass';
postgres=# GRANT ALL PRIVILEGES ON DATABASE mydb TO testuser;
```

With the correct params to connect to the created PG database, the module will automatically creates two tables (if they do not exist): `hauth_user` and `hauth_role`.

## Config parameters

Module `hauth` takes in two sets of parameters:

```javascript
const hauth = require('@horanet/hauth');
hauth.init(params, db);
```

where `db` is a database handle, for example

```js
const db = new pg.Pool({
  host: 'localhost',
  port: 5432,
  database: 'test',
  user: 'testuser',
  password: 'testpass'
});
```

and `params` is an object containing the configuration parameters (all optional) :

### `cookiename`

the default value is `hauth`

### JWT parameters

as defined in [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken) module

- `jwt_key`: key used to cipher JSON web tokens, optional ; if not defined,randomly generated at each server startup
- `jwt_alg`: algorithm used for signing JWT, default value is `HS256`
- `jwt_exp`: expiration time for JWT, default `2h` for 2 hours

### `roles`

an array with the name of the roles. Each user can be assigned one role granting access to some URLs according to `accessRules`. For example, `roles = ['admin', 'basic_user']`

### `accessRules`

dictate the behavior of the control function, according to the URLs, with the following format:

```config
<pattern>: <keyword> or <list of accepted profiles> 
```

where `pattern` is:

- a path, matching the starting of the request path (hence, starting with `/`), e.g. `'/files/'`
- or a [regex](https://en.wikipedia.org/wiki/Regular_expression) string matching a part of the request path, e.g. `'\.css$'` (a pattern is recognized as a regex by the fact that it does not start with `/`)

The keywords are:

- `allow`: allowed access to any authenticated user
- `deny`: access forbidden
- `skip`: allowed access without authentication

These rules are tested in the order they are listed, until a rule matches.

If no access rule matches the URL, the default rule is `allow`. So, in order to forbid anything that is not explicitly allowed, you should end the access rules with the rule `'/': 'deny'`

### `on401`

A function to provide custom response if user authentication is required (typically used to send an authentication form)

### `on403`

A function to provide custom response in case of access denied

### `onLogout`

A function to provide custom response when the user logs out

### `autocreate`

A function to make possible for an unregistered client to create its own accounts. If this function is defined, if an unknown client provides username and password in an 'Authorization: Basic' header, it makes it possible to check the credentials and to run some more code; if this function returns a non-null object describing an account, Hauth will create the account in hauth_user and deliver a new password (in a 'X-Next-Password' Header).

For example,

```js
  autocreate: async function(login, pwd, db) {
    if (pwd === 'secret') { // this magic password allows to create any admin account
      await db.query('INSERT INTO ...'); // you can add some extra processing
      return {login: login, name: login, role: 'admin'}
    }
  }
```

### `defaultUsers`

An array of users to be created along with tables hauth_users and hauth_roles. This is intended for initing apps; These accounts can then be modified or deleted, and will not be recreated at each server startup.

## APIs

This `hauth` module exposes the following functions:

### init

Initialize the module. This function takes in the `config` and `db` parameters ; it creates the two tables `hauth_role` and `hauth_user` if necessary (and thus fills the default users), then fill in the `roles` defined in `config`.

### control

Runs authentication and access control

### allowed

Function allowed(role, url): this function return `True` if a user with rule `<rule>` is authorized to access `<url>`.

### getCookie

Sign and verify token using `jsonwebtoken` module.

### delCookie

Delete cookie when the current user logs out.

### addUser

Add a single user to table `hauth_user`

- Input parameter `user`: a JSON object containing user's information. For example:

```config
{"login": "admin", "name": "Administrator", "role": "admin", "password": "admin"}
```

- Not all information are required, if a required field is missing, default information will be filled.

### delUser

Delete a user from table `hauth_user`:

- Input parameter `user`: a JSON object containing user's information
- Require field: `user.login`
- User with `login` will be deleted
- Return `pass` value: `true` in case the `DELETE` query was succesfull, `false` otherwise.

### modUser

Modify existing user in table `hauth_user`:

- Input parameter `user`: a JSON object containing user's information
- Require field: `user.login`
- Information of user with `login` will be updated according to the input fields
- Return `pass` value: `true` in case the `UPDATE` query was succesfull, `false` otherwise.

## Example

To run the example (the script located in `./example/index.js`):

```shell
npm install
npm run example
```

To check the currently logged in user, use the path `localhost:3000/whoami`

## Test

## FAQ


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.5](https://gitlab.horanet.com/hframework/hauth/compare/v0.1.4...v0.1.5) (2021-10-13)


### Features

* config parameters can be changed after init ([a029201](https://gitlab.horanet.com/hframework/hauth/commit/a029201b8387727f05226ddbe6ca1b7c95dae01e))

### [0.1.4](https://gitlab.horanet.com/hframework/hauth/compare/v0.1.3...v0.1.4) (2021-09-08)


### Features

* export function getUser ([aa31e00](https://gitlab.horanet.com/hframework/hauth/commit/aa31e0043899cb1c73f89b9f2837be8fc918b49b))

### [0.1.4](https://gitlab.horanet.com/hframework/hauth/compare/v0.1.3...v0.1.4) (2021-09-08)


### Features

* export function getUser ([aa31e00](https://gitlab.horanet.com/hframework/hauth/commit/aa31e0043899cb1c73f89b9f2837be8fc918b49b))

### [0.1.3](https://gitlab.horanet.com/hframework/hauth/compare/v0.1.2...v0.1.3) (2021-09-03)


### Bug Fixes

* Excessive package size ([b5af6c0](https://gitlab.horanet.com/hframework/hauth/commit/b5af6c01e0e4343ba8089bcf622f159835db4d44))

### [0.1.2](https://gitlab.horanet.com/hframework/hauth/compare/v0.1.1...v0.1.2) (2021-09-03)


### Features

* add param onLogout ([d0862a1](https://gitlab.horanet.com/hframework/hauth/commit/d0862a10f0adc9ac51e3041f8f31f63eb49c42de))

### 0.1.1 (2021-08-25)


### Bug Fixes

* md file linting (revert No of version back to 0.1.0, this will be changed by npm run release) ([a55b13a](https://gitlab.horanet.com/hframework/hauth/commit/a55b13acab50fe9aeb15fb389e165891a149d77f))

### [0.0.6](https://gitlab.horanet.com/hframework/hauth/compare/v0.0.5...v0.0.6) (2021-07-09)


### Bug Fixes

* modify example for version release ([d02bb86](https://gitlab.horanet.com/hframework/hauth/commit/d02bb863d3ce519abab85daa557150b43a53802f))

### [0.0.5](https://gitlab.horanet.com/hframework/hauth/compare/v0.0.4...v0.0.5) (2021-06-11)

### [0.0.4](https://github.com/naml3i/hauth-dev/compare/v0.0.3...v0.0.4) (2021-06-02)

### 0.0.2 (2021-06-01)

- Change author name in git repo to match name in other machines.
- Testing automatic versioning with standard-version (instead of npm version)


lib/index.js
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

var defCfg = { // config parameters, populated with default values
    cookiename: 'hauth',
    jwt_key: pwgen(), // key used for signing JWT: if not defined,
                      // it is randomly generated at each server startup
    jwt_alg: 'HS256', // algorithm used for signing JWT
    jwt_exp: '2h',    // expiration time for JWT, default 2 hours
};
var extCfg = {};
var db;

function checkInit(fun) {
  return (...args) => {
    if (!db) throw('Hauth is not inited');
    return fun(...args);
  }
}
module.exports = {
    init: init,
    control:   checkInit(control),
    allowed:   checkInit(allowed),
    getCookie: checkInit(getCookie),
    delCookie: checkInit(delCookie),
    getUser:   checkInit(getUser),
    addUser:   checkInit(addUser),
    delUser:   checkInit(delUser),
    modUser:   checkInit(modUser),
    addRoles:  checkInit(addRoles),
};

async function init (config, dbh) {
  if (db) {
    console.warn('Hauth is already inited');
    return;
  }
  db = dbh;
  extCfg = config;

  await checkTableRole();
  await addRoles(cfg().roles);
  await checkTableUser();
}

function cfg() {
  return { ...defCfg, ...extCfg };
}

/**
 * check if a table exists in database
 */
async function missing (tablename) {
  const test = await db.query(
    `SELECT count(*) FROM pg_catalog.pg_tables WHERE tablename=$1`,
    [tablename]
  );
  return test.rows[0].count === '0';
}

/**
 * check if table 'hauth_role' exists, create if missing
 */
async function checkTableRole() {
  if (await missing('hauth_role')) {
    console.info('Creating table hauth_role')
    await db.query(`
      CREATE TABLE hauth_role (
        id SERIAL PRIMARY KEY,
        name VARCHAR(20) NOT NULL,
        UNIQUE (name)
      )`
    );
  }
}

/**
 * check if table 'hauth_user' exists, create if missing
 */
async function checkTableUser() {
  if (await missing('hauth_user')) {
    console.info('Creating table hauth_user')
    await db.query(`
      CREATE TABLE hauth_user (
        id SERIAL PRIMARY KEY,
        login VARCHAR(50) NOT NULL,
        name VARCHAR(100),
        role_id INTEGER REFERENCES hauth_role (id) ON DELETE SET NULL ON UPDATE CASCADE,
        password VARCHAR(100),
        next_password VARCHAR(100),
        UNIQUE (login)
      )`
    );
    for (const user of cfg().defaultUsers) {
      addUser(user);
    }
  }
}

/**
 * Populate table hauth_role with roles
 * @param {*} roles: as defined in the config. E.g. ['admin', 'installer']
 */
async function addRoles(roles) {
  const existingRoles = await db.query(`SELECT name FROM hauth_role`);
  newRoles = roles.filter( (role) => !existingRoles.rows.map((x) => x.name).includes(role) );
  if (newRoles.length) {
    var i=1;
    const dollars = newRoles.map((x) => `($${i++})`).join(',');
    await db.query(`INSERT INTO hauth_role (name) VALUES ${dollars} ON CONFLICT DO NOTHING`, newRoles);
  }
}

async function parse(user) {
  const role = user.role;
  delete user.role;

  if (user.password) {
    user.password = await bcrypt.hash(user.password, 10);
  }

  var cols = Object.keys(user);
  var vals = Object.values(user);
  var i = 0;
  var dols = vals.map((x) => `$${++i}`);

  if (role) {
    cols.push('role_id');
    dols.push(`(SELECT id FROM hauth_role WHERE name=$${++i})`);
    vals.push(role);
  }
  const res = [cols.join(','), dols.join(','), vals, i];
  return [cols.join(','), dols.join(','), vals, i];
}

/**
 * Add a single user to the hauth_user
 * @param {*} user JSON object
 * @returns the new user profile if insert is successful, else null
 */
async function addUser(user) {
  var [cols, dols, vals, i] = await parse(user);
  const result = await db.query(`INSERT INTO hauth_user(${cols}) VALUES (${dols}) ON CONFLICT DO NOTHING RETURNING *`, vals);
  return result.rowCount > 0 ? result.rows[0] : null;
}

/**
 * Modify existing user
 * @param {*} login : user's or device's login ID
 * @param {*} user : object containing the fields to be updated
 * @returns the new user profile if update is successful, else null
 */
async function modUser(login, user) {
  var [cols, dols, vals, i] = await parse(user);
  if (!i) return false;
  if (i > 1) { [cols, dols] = [`(${cols})`, `(${dols})`]; }
  const result = await db.query(`UPDATE hauth_user SET ${cols} = ${dols} WHERE login=$${++i} RETURNING *`, [...vals, login]);
  return result.rowCount > 0 ? result.rows[0] : null;
}

/**
 * Delete existing user
 * @param {*} user 
 * @returns the removed user profile if delete is successful, else null
 */
async function delUser(login) {
  const result = await db.query(`DELETE FROM hauth_user WHERE login=$1 RETURNING *`, [login]);
  return result.rowCount > 0 ? result.rows[0] : null;
}

/**
 * authentication and access control (display page/error page...)
 * based on access rule of each path user
 */
async function control (req, res, next) {
  const rule = getRule(req.path);
  if (rule === 'skip') {
    next();
  } else {
    if (req.user || checkToken(req) || await checkUser(req, res)) {
      if (allowed(req.user.role, req.path, rule)) {
        next();
      } else {
        res.status(403);
        if (cfg().on403) {
          cfg().on403(req, res);
        } else {
          res.send();
        }
      }
    } else {
      res.status(401)
      if (cfg().on401) {
        cfg().on401(req, res);
      } else {
        res.send();
      }
    }
  }
}

/**
 * returns access rule for the path, based on config
 */
function getRule(url) {
  for (const [pattern, rule] of Object.entries(cfg().accessRules)) {
    if (pattern.startsWith('/') && url.startsWith(pattern) || new RegExp(pattern).test(url)) {
      return rule
    }
  }
  return 'allow';
}

/**
 * @param {*} role 
 * @param {*} url 
 * @param {*} rule optional: to avoid rechecking the access rule if it's already checked
 * @returns true if a user of role <role> is authorized to access <url>
 */
function allowed(role, url, rule) {
  rule = rule || getRule(url);
  if (!rule || rule === 'deny') {
    return false;
  } else if (rule === 'allow' || rule === 'skip') {
    return true;
  } else if (Array.isArray(rule)) {
    return rule.includes(role) ? true : false;
  }
  console.error(`invalid access rule "${rule}" in hauth config`);
  return false;
}

/**
 * Validate the JWT sent in the cookie
 * @param {*} req HTTP request message
 * @returns true if the JWT is verified
 */
function checkToken(req) {
  const token = req.cookies[cfg().cookiename];
  if (token) {
    try {
      req.user = jwt.verify(token, cfg().jwt_key, {algorithms: cfg().jwt_alg});
      return true;
    } catch (error) {
    }
  }
}

/**
 * Verify the credential submitted
 * @returns true or false 
 */
async function checkUser(req, res) {
  const [login, pwd] = getUserAndPwdFromReq(req);
  if (login && pwd !== undefined) { // le mot de passe peut être ''
    var [user, dbPwd, nextPwd] = await getUser(login);
    if (user) {
      if (pwd === nextPwd) {
        req.user = user;
        await updatePwd(login, pwd);
        return true;
      }
      if (dbPwd && (pwd === dbPwd || await bcrypt.compare(pwd, dbPwd))) {
        req.user = user;
        if (nextPwd) {
          res.set('X-Next-Password', nextPwd);
        }
        return true;
      }
      console.warn(`Attempt to login as ${login} with bad password`);
    } else { // login inconnu
      if (cfg().autocreate) {
        user = await cfg().autocreate(login, pwd, db);
        if (user) {
          req.user = user;
          user.next_password = pwgen(); // génération d'un mot de passe
          res.set('X-Next-Password', user.next_password);
          await addUser({...user});
          return true;
        }
      }
      console.warn(`Attempt to login as ${login}, but this login is unknown`)
    }
  }
  return false;
}

/** fonction getUserAndPwdFromReq :
 * extrait login et mot de passe, soit du corps de la requête,
 * soit de l'en-tête de requête HTTP 'Authorization: Basic xxx'
 */
function getUserAndPwdFromReq(req) {
  const authz = req.headers.authorization;
  if (authz && authz.startsWith('Basic ')) {
    const value = Buffer.from(authz.split(' ')[1], 'base64').toString('utf-8');
    return value.includes(':') ? value.split(':') : [];
  } else {
    return req.body ? [req.body.login, req.body.password] : [];
  }
}

/**
 * Extract user, password and next_password (if applicable) from hauth_user
 * @param {*} login: user's login ID
 * @returns 
 */
async function getUser(login) {
  const query = "SELECT hauth_user.*, hauth_role.name as role FROM hauth_user LEFT JOIN hauth_role ON role_id=hauth_role.id WHERE login=$1";
  const users = await db.query(query, [login]);
  if (users.rowCount) {
    const user = users.rows[0];
    const [pwd, nextPwd] = [user.password, user.next_password];
    delete user.password;
    delete user.next_password;
    delete user.role_id;
    return [user, pwd, nextPwd];
  } else {
    return [];
  }
}

async function updatePwd(login, pwd) {
  const hash = await bcrypt.hash(pwd, 10);
  await db.query(`UPDATE hauth_user SET password=$2, next_password=NULL WHERE login=$1`, [login, hash]);
}

/** 
 * @returns a pseudo-random password of 10 characters [0-9 a-z]
 */
function pwgen() {
  return Math.random().toString(36).substring(2, 12);
}

async function getCookie(req, res) {
  res.set('Cache-Control', 'no-cache, private, no-store, must-revalidate');
  if (await checkUser(req, res)) {
    const token = jwt.sign(req.user, cfg().jwt_key, {algorithm: cfg().jwt_alg, expiresIn: cfg().jwt_exp});
    res.cookie(cfg().cookiename, token, {httpOnly: true});
    res.status(200).send(req.user);
  } else {
    res.status(401).send();
  }
};

function delCookie(req, res) {
  res.clearCookie(cfg().cookiename);
  if (cfg().onLogout) {
      cfg().onLogout(req, res);
  } else {
      res.send();
  }
};

